sklearn/tree/_tree.pyx

```diff
@@ -2693,15 +3403,16 @@ cdef class DepthFirstTreeBuilder(TreeBuilder):
                            (n_node_samples < 2 * min_samples_leaf) or
                            (weighted_n_node_samples < min_weight_leaf))
 
+                if not is_leaf:
+                    splitter.node_split(impurity, &split, &n_constant_features)
+                    is_leaf = is_leaf or (split.pos >= end)
+
                 if first:
                     impurity = splitter.node_impurity()
                     first = 0
 
                 is_leaf = is_leaf or (impurity <= MIN_IMPURITY_SPLIT)
 
-                if not is_leaf:
-                    splitter.node_split(impurity, &split, &n_constant_features)
-                    is_leaf = is_leaf or (split.pos >= end)
```

Comment:
The code was organized such that the impurity of the entire dataset was calculated before the first split. However, when you calculate the split, you also get this value. My code caches this value when it does the split, but needs to do the split first before calculating the impurity.

This has been updated in the latest commits, but only through a hack. I am now working on making this cleaner.